tests/testsuite/doc.rs

```diff
@@ -2349,3 +2349,24 @@ fn scrape_examples_missing_flag() {
         .with_stderr("error: -Z rustdoc-scrape-examples must take [..] an argument")
         .run();
 }
+
+#[cargo_test]
+fn link_to_private_item() {
+    let main = r#"
+    //! [foo]
+    fn foo() {}
+    fn main() {
+        foo();
+    }
+    "#;
+    let p = project()
+        .file("src/lib.rs", main)
+        .file("src/main.rs", main)
+        .build();
+    p.cargo("doc --lib")
+        .with_stderr_contains("[..] links to private item `foo`")
+        .run();
+    p.cargo("doc --bin foo")
+        .with_stderr_does_not_contain("[..] links to private item `foo`")
+        .run();
```

Comment:
I'm not sure this test is checking for what this PR is trying to resolve. It passes whether or not `-Arustdoc::link_to_private_item` is included.  The output here appears to be:

```
warning: function is never used: `foo`
 --> src/lib.rs:2:8
  |
2 |     fn foo() {}
  |        ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function is never used: `main`
 --> src/lib.rs:3:8
  |
3 |     fn main() {
  |        ^^^^

warning: `foo` (lib) generated 2 warnings
warning: `foo` is both a crate and a function
 --> src/main.rs:1:10
  |
1 |     //! [foo]
  |          ^^^ ambiguous link
  |
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
help: to link to the module, prefix with `mod@`
  |
1 |     //! [mod@foo]
  |          ++++
help: to link to the function, add parentheses
  |
1 |     //! [foo()]
  |             ++

warning: `foo` (bin "foo" doc) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
```

Perhaps it would be better to use `with_stderr` to verify the entire output. `does_not_contain` is generally a fragile thing to use in tests.

I might suggest using different examples for `lib.rs` and `main.rs`, and don't use `[foo]`, since it is ambiguous.